### ISSUE TITLE
[Proposal] Method & constructor overloading

### ISSUE THREAD

**Issue by MaxGraey** [AssemblyScript/assemblyscript#816]
This little bit verbose but valid TypeScript syntax for constructor overloading. Will be nice to have this

```ts
class Field {
  public name: string;
  public surname: string;

  constructor();
  constructor(name: string);
  constructor(name: string, surname: string);
  constructor(name?: string, surname?: string) {
    if (name && surname) {
      this.name    = name;
      this.surname = surname;
    } else if (name && !surname) {
      this.name    = name;
      this.surname = "<unknown>";
    } else {
      this.name    = "<unknown>";
      this.surname = "<unknown>";
    }
  }
}

console.log(new Field('Jonh','Smith')); // Field { name: 'Jonh', surname: 'Smith' }
console.log(new Field('Jonh')); // Field { name: 'Jonh', surname: '<unknown>' }
console.log(new Field()); // Field { name: '<unknown>', surname: '<unknown>' }
```

**Comment by MaxGraey** [comment]
This feature unblock some standard library methods like `Array.from` with different signatures

**Comment by MaxGraey** [comment]
Currently we can partially emulate this via generics but this not intellisense-friendly and not possible for constructors

**Comment by willemneal** [comment]
I know that we want to try to match TS as much as possible, especially because the IDE support, but one area where we can be an improvement over TS is with proper method overloading.  When there is still an intersection of the types used at the call site and the various signatures, e.g. here it would be `new Field("John", "Smith");` which could be the second or the forth constructor.  However, in the cases where they are disjoint, e.g. (assuming that the last constructor only took a string) `new Field(123)` could be it's own function.

```ts
class Field {
  public name: string = "<unknown>";
  public surname: string = "<unknown>";

  constructor() {}
  constructor(name: string) {
    this.name = name;
  }
  constructor(name: string, surname: string = "<unknown>") {
    this.name = name;
    this.surname = surname;
  }
  constructor(name: i32, surname: string = "<unknown>") {
    this.name = name.toString();
    this.surname = surname;
  }  
}
```

I do understand the downside to moving away from complete TS support in IDE's, but in my opinion the above is much clearer and we could even allow for them to call each other. e.g.

```ts
class Field {
  public name: string = "<unknown>";
  public surname: string = "<unknown>";

  constructor() {}
  constructor(name: string) {
    this.name = name;
  }
  constructor(name: string, surname?: string) {
    constructor(name);
    if (!!surname){
      this.surname = surname;
    }
  }
  constructor(name: i32, surname?: string) {
    constructor(name.toString(), surname);
  }  
}
```

The benefit here is that they could be used in interfaces this way too.

**Comment by dcodeIO** [comment]
Maybe we can make breaking with TS optional. So if someone wants it, they can do that, or if they want compatibility, they'd use TS's notation and accept the limitations?

**Comment by MaxGraey** [comment]
@dcodeIO yeah, makes sense.

@willemneal Great! Just several suggestions
```ts
class Field {
  public name: string = "<unknown>";
  public surname: string = "<unknown>";

  constructor() {}

  constructor(name: string) {
    this.name = name;
  }

  constructor(name: string, surname?: string) {
    this(name); // or this.constructor ?
    if (!!surname) {
      this.surname = surname;
    }
  }

  constructor(name: i32, surname?: string) {
    this(name.toString(), surname);
  }  
}
```

Which will be more consistent with `super()`

**Comment by willemneal** [comment]
Thanks!  `this()` is what Java and C# do, but `this.constructor()` is clearer.  Perhaps stick with `this` since users that are comfortable with constructor overloading would expect it.

**Comment by DanielRX** [comment]
> @dcodeIO yeah, makes sense.
> 
> @willemneal Great! Just one suggestion
> 
> ```ts
> class Field {
>   ...
> }
> ```
> 
> Which will be more consistent with `super()`

Wouldn't
>   ```
>    constructor(name: string, surname?: string) {
>     this(name); // or this.constructor ?
>     if (surname) this.surname = surname;
>   }
>    ```
be
```
constructor(name: string, surname: string) {
    this(name); // or this.constructor ?
    this.surname = surname;
}
```
since as it stands, this can already be merged with the existing constructor (without surname)?

**Comment by MaxGraey** [comment]
So as I promised on recent meeting I suggest new approach for overloading methods which partially compatible with TS.

Example:
```ts
import utils from './utils';

export class Color {
    static fromNum(color: number): Color {
        return new Color(
            (color >> 24) & 0xFF,
            (color >> 16) & 0xFF,
            (color >>  8) & 0xFF,
            (color >>  0) & 0xFF
        );
    }

    static fromStr(color: string): Color {
        return new Color(utils.parse(color));
    }

    @overload(fromNum) constructor(color: number);
    @overload(fromStr) constructor(color: string);
    constructor(a: number, r: number, g: number, b: number);
    constructor() {
        // default implementation
    }
}
```

Thoughts?

**Comment by dcodeIO** [comment]
Might work, but looks quite complicated. Personally I'd most likely do something more reliable like this:

```ts
export class Color {
    static create<T>(value: T): Color {
      if (isInteger<T>()) { ... }
      else if (isString<T>()) { ... }
      else ERROR("...");
    }
    constructor(a: u32, r: u32, g: u32, b: u32)
        this.a = a; this.r = r;  this.g = g; this.b = b;
    }
}
```

**Comment by MaxGraey** [comment]
Yes, but what if you want different implementation for different arity (number of args)?

**Comment by dcodeIO** [comment]
Ah, sorry, covered this in an edit :)

**Comment by MaxGraey** [comment]
```ts
class Color {
  constructor(a: u32, r: u32, g: u32, b: u32)
    this.a = a; this.r = r;  this.g = g; this.b = b;
  }
  constructor(s: string) {
    this(...);
  }
}
```
Will be ideal but it completely uncompatible with TS. Perhaps with our own language server it will be possible, but we loose potability mode

**Comment by dcodeIO** [comment]
There's also the option to support union types in scenarios like this (just not for variables), compiling them similar to generics:

```ts
  constructor(a: string | u32, r: u32 = 0, g: u32 = 0, b: u32 = 0)
    if (isString(a)) {
      ...
    } else {
      this.a = a; this.r = r;  this.g = g; this.b = b;
    }
  }
```

giving us a `constructor<string,u32,u32,u32>` and a `constructor<u32,u32,u32,u32>` depending on arguments.

**Comment by MaxGraey** [comment]
Yeah, but it much more complicated and potentially error-prone approach

**Comment by dcodeIO** [comment]
Perhaps we also can

```ts
export class Color {
    constructor(s: string);
    constructor(a: u32, r: u32, g: u32, b: u32);
    constructor() {
        // inject arguments[0..3] and arguments.length statically?
        // so either arguments.length == 1 and arguments[0] -> string
        // or arguments.length == 4 and arguments[0..3] -> u32
    }
}
```

**Comment by MaxGraey** [comment]
Yeah. And in this case all checks better do statically. That's most TypeScrip-tish way

**Comment by trusktr** [comment]
> this(...);

That's only compatible with extending `Function`. Can we have an option to make any non-TS/JS-compatible syntax be a compile error? I think that'd be great!

**Comment by MaxGraey** [comment]
Another possible alternative:
```ts
export class Color {
    @overload((s: string) => this);
    constructor(a: u32, r: u32, g: u32, b: u32) {
    }

    @overload((color: string) => void)
    setColor(a: u32, r: u32, g: u32, b: u32): void {
        // @ts-ignore
        if (color instanceof string) {
            // parse from string
        } else {
            // use a, r, g, b
        }
    }
}
```

**Comment by munrocket** [comment]
Constructor overloading can be very useful if your compilation target is only webassembly. Obviously that some switch will be better for those who don't want to compile sources to javascript. Sometime this is very handy because the language itself become more powerful.

**Comment by trusktr** [comment]
I think AS just needs unions. Then using TypeScript overload syntax will be relatively easy.

Also, supporting multiple syntaxes would make more maintainance burden, right?

**Comment by aminya** [comment]
Whichever approach is taken it should be compile time dispatch. Using if/else means it is runtime dispatch which does not have any benefits.

Regarding the syntax, I suggest breaking TS syntax here which is in my opinion very ugly and strange.

**Comment by trusktr** [comment]
> Using if/else means it is runtime dispatch which does not have any benefits.

Note that with strict types, conditional branches are eliminated during compile time based on what types the compiler detects are used at the call sites.

> Regarding the syntax, I suggest breaking TS syntax here which is in my opinion very ugly and strange.

We should not break from TypeScript syntax (in any way we already have, we should provide compatible alternatives). That's a major benefit of AssemblyScript (the wide array of TypeScript tooling that is already available).

**Comment by trusktr** [comment]
> Maybe we can make breaking with TS optional. So if someone wants it, they can do that, or if they want compatibility, they'd use TS's notation and accept the limitations?

This may create fragmentation:

It may be nice for authors to choose how they write (and for every incompatible syntax we'd expect there to be a compatible syntax option too), but what happens when downstream library users import the code into their environments?

Suppose a library user wants compatibility with tooling and they write code using compatible features. What happens when they import a library that doesn't us compatible features? If this doesn't work, it will cause fragmentation.

**Comment by trusktr** [comment]
I know that `a * b` is a tiny bit more convenient than `a.times(b)`, but in the grand scheme of things, looking at the big picture, the less concise syntax is a very *tiny* price to pay for the big benefits that compatibility brings.

**Comment by aminya** [comment]
Well, the description of AssemblyScript is "Definitely not a TypeScript to WebAssembly compiler". Trying to stay compatible with TypeScript's handling of functions/methods that is designed for JavaScript is just continuing a mistake in the design. I am all for backward compatibility as long as it is not a technical debt. AssemblyScript is young. It should not limit itself to something that was decided in 1995 for JavaScript.
